### DISCUSSION TITLE
A way to tracking errors

### DISCUSSION THREAD

**Discussion by alihassan161820** [directus/directus#11475] [Archive]
I tried to integrate with sentry in init hooks to track 500 errors that printed in console but it didn't catch any error so is there any hook to catch threw  errors in it ?

```js
module.exports = function ({ init }, { services, env, database }) {
  init("routes.before", async (app) => {
    Sentry.init({
      dsn: "https://------------------------------------------@o1037532.ingest.sentry.io/6005525",
    });
    app.use(Sentry.Handlers.requestHandler());
    app.use(Sentry.Handlers.errorHandler());
  });
};
```

<details>
<summary>More Info</summary>

### What version of Directus are you using?

9.3.1

### What version of Node.js are you using?

16

### What database are you using?

mysql

### What browser are you using?

chrome

### What operating system are you using?

ubuntu

### How are you deploying Directus?

docker

</details>

**Comment by azrikahar** [comment]
Converted this to discussion since it's more of a question rather than a bug 👍

---

With the disclaimer that I don't really have experience with Sentry personally, based on what I read over here: https://docs.sentry.io/platforms/node/guides/express/

It does mention that error handler should be _after_ all controllers:

```js
// The error handler must be before any other error middleware and after all controllers
app.use(Sentry.Handlers.errorHandler());
```

Whereas since you are currently using `routes.before`, that is effectively before all the registered routes internally, as seen in the source:

https://github.com/directus/directus/blob/b8c4317cb1cd9a8ed0eeb1ad4db35f8771588624/api/src/app.ts#L199-L237

With the advice by Sentry's documentation in mind, perhaps you can try putting `app.use(Sentry.Handlers.errorHandler());` inside `routes.custom.after` instead and see how it goes.

**Comment by unleashit** [comment]
This is a bit old, but in case it helps anyone else after some tinkering I was able to get it to work. Sentry requires two middlewares, one before all others and one at the end (but before the main error mw). The hooks that did the trick are `routes.before` and `routes.custom.after`.

```javascript
import * as Sentry from "@sentry/node";

export default ({ init }, { env }) => {
	const { SENTRY_DSN } = env;

	Sentry.init({
		dsn: SENTRY_DSN,
	});

	init('routes.before', ({ app }) => {
		app.use(Sentry.Handlers.requestHandler());
		console.log('-- Sentry Request Handler Added --');
	}));

	init('routes.custom.after', ({ app }) => {
		app.use(Sentry.Handlers.errorHandler());
		console.log('-- Sentry Error Handler Added --');
	});
};
```
Sidenote, I tried at first to set up a Flow to generate an error for testing... but for some reason my errors were swallowed (not even showing in Directus logs). Not sure if this is intentional or a bug. In any case, I was able to create an endpoint and test that way.
